src/cloud-api-adaptor/docs/addnewprovider.md

```diff
@@ -53,10 +53,371 @@ To include your provider you need reference it from the main package. Go build t
+
+[[ -S ${CRI_RUNTIME_ENDPOINT} ]] && optionals+="-cri-runtime-endpoint ${CRI_RUNTIME_ENDPOINT} "
+[[ "${PAUSE_IMAGE}" ]] && optionals+="-pause-image ${PAUSE_IMAGE} "
+[[ "${VXLAN_PORT}" ]] && optionals+="-vxlan-port ${VXLAN_PORT} "
+[[ "${CACERT_FILE}" ]] && optionals+="-ca-cert-file ${CACERT_FILE} "
+[[ "${CERT_FILE}" ]] && [[ "${CERT_KEY}" ]] && optionals+="-cert-file ${CERT_FILE} -cert-key ${CERT_KEY} "
+[[ "${TLS_SKIP_VERIFY}" ]] && optionals+="-tls-skip-verify "
+[[ "${PROXY_TIMEOUT}" ]] && optionals+="-proxy-timeout ${PROXY_TIMEOUT} "
+[[ "${AA_KBC_PARAMS}" ]] && optionals+="-aa-kbc-params ${AA_KBC_PARAMS} "
+[[ "${FORWARDER_PORT}" ]] && optionals+="-forwarder-port ${FORWARDER_PORT} "
+[[ "${CLOUD_CONFIG_VERIFY}" == "true" ]] && optionals+="-cloud-config-verify "
+
+test_vars() {
+    for i in "$@"; do
+        [ -z "${!i}" ] && echo "\$$i is NOT set" && EXT=1
+    done
+    [[ -n $EXT ]] && exit 1
+}
+
+one_of() {
+    for i in "$@"; do
+        [ -n "${!i}" ] && echo "\$$i is SET" && EXIST=1
+    done
+    [[ -z $EXIST ]] && echo "At least one of these must be SET: $*" && exit 1
+}
+
+libvirt() {
+    test_vars LIBVIRT_URI
+
+    [[ "${DISABLECVM}" = "true" ]] && optionals+="-disable-cvm "
+    set -x
+    exec cloud-api-adaptor libvirt \
+        -uri "${LIBVIRT_URI}" \
+        -data-dir /opt/data-dir \
+        -pods-dir /run/peerpod/pods \
+        -network-name "${LIBVIRT_NET:-default}" \
+        -pool-name "${LIBVIRT_POOL:-default}" \
+        ${optionals} \
+        -socket /run/peerpod/hypervisor.sock
+}
+
+help_msg() {
+    cat <<EOF
+Usage:
+	CLOUD_PROVIDER=libvirt $0
+or
+	$0 libvirt
+in addition all cloud provider specific env variables must be set and valid
+(CLOUD_PROVIDER is currently set to "$CLOUD_PROVIDER")
+EOF
+}
+
+if [[ "$CLOUD_PROVIDER" == "libvirt" ]]; then
+    libvirt
+else
+    help_msg
+fi
+```
+### Step 8: Update cloud-api-adaptor damonset to use the external `libvirt` plugin
+- Overwrite the command
+```
+    spec:
+      containers:
+      - command:
+        - /run/cloud-api-adaptor/plugins/entrypoint.sh
+```
+- Mout `/run/cloud-api-adaptor/plugins/` from worker node to the `cloud-api-adaptor-con` container
+```yaml
+...
+        volumeMounts:
+        - mountPath: /run/cloud-api-adaptor/plugins
+          name: provider-dir
+...
+      volumes:
+      - hostPath:
+          path: /run/cloud-api-adaptor/plugins
+          type: Directory
+        name: provider-dir
+...
+```
+### Step 9: Update peerpod-ctrl deploy to use the external `libvirt` plugin
+- Mount `/run/cloud-api-adaptor/plugins` from worker node to the `manager` container
+```yaml
+...
+      volumes:
+      - hostPath:
+          path: /run/cloud-api-adaptor/plugins
+          type: Directory
+        name: provider-dir
+...
+        volumeMounts:
+        - mountPath: /run/cloud-api-adaptor/plugins
+          name: provider-dir
+...
+```
+### Step 10: Verify cloud-api-adaptor/peerpod-ctrl pod is running without error
```

Comment:
On the worker I corrected the permissions of the entrypoint.sh:
```
root@peer-pods-worker-0:~# ls -al /run/cloud-api-adaptor/plugins/entrypoint.sh
-rwxr-xr-x 1 root root 2329 Apr 16 14:57 /run/cloud-api-adaptor/plugins/entrypoint.sh
```
but the pod still failed with a permissions error:
```
# kubectl describe pod -lapp=cloud-api-adaptor -n confidential-containers-system
Name:             cloud-api-adaptor-daemonset-js6rh
Namespace:        confidential-containers-system
Priority:         0
Service Account:  cloud-api-adaptor
Node:             peer-pods-worker-0/192.168.122.38
Start Time:       Tue, 16 Apr 2024 16:58:56 +0000
Labels:           app=cloud-api-adaptor
                  controller-revision-hash=76f7b78b78
                  pod-template-generation=3
Annotations:      <none>
Status:           Running
IP:               192.168.122.38
IPs:
  IP:           192.168.122.38
Controlled By:  DaemonSet/cloud-api-adaptor-daemonset
Containers:
  cloud-api-adaptor-con:
    Container ID:  containerd://2e5cc8073fa73e07db74b66e7d2e42ea0338c40cf98009a30dab902f8293bead
    Image:         quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec
    Image ID:      quay.io/stevenhorsman/cloud-api-adaptor@sha256:f7093ead38b8274517739791609f0e62f317a6005b8cb8824eae2e01a3775138
    Port:          <none>
    Host Port:     <none>
    Command:
      /run/cloud-api-adaptor/plugins/entrypoint.sh
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       StartError
      Message:      failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: eaccess /run/cloud-api-adaptor/plugins/entrypoint.sh: permission denied: unknown
      Exit Code:    128
      Started:      Thu, 01 Jan 1970 00:00:00 +0000
      Finished:     Tue, 16 Apr 2024 16:59:45 +0000
    Ready:          False
    Restart Count:  3
    Startup:        http-get http://:8000/startup delay=20s timeout=1s period=20s #success=1 #failure=30
    Environment Variables from:
      peer-pods-secret  Secret     Optional: false
      peer-pods-cm      ConfigMap  Optional: false
    Environment:
      NODE_NAME:   (v1:spec.nodeName)
    Mounts:
      /root/.ssh/ from ssh (ro)
      /root/containers/ from auth-json (ro)
      /run/cloud-api-adaptor/plugins from provider-dir (rw)
      /run/netns from netns (rw)
      /run/peerpod from pods-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-thtb6 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  auth-json:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  auth-json-secret
    Optional:    true
  ssh:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ssh-key-secret
    Optional:    true
  pods-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /run/peerpod
    HostPathType:
  netns:
    Type:          HostPath (bare host directory volume)
    Path:          /run/netns
    HostPathType:
  provider-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /run/cloud-api-adaptor/plugins
    HostPathType:  Directory
  kube-api-access-thtb6:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              node.kubernetes.io/worker=
Tolerations:                 node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/network-unavailable:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  78s                default-scheduler  Successfully assigned confidential-containers-system/cloud-api-adaptor-daemonset-js6rh to peer-pods-worker-0
  Normal   Pulled     77s                kubelet            Successfully pulled image "quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec" in 609.593825ms (609.657269ms including waiting)
  Normal   Pulled     75s                kubelet            Successfully pulled image "quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec" in 581.294288ms (581.324452ms including waiting)
  Normal   Pulled     61s                kubelet            Successfully pulled image "quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec" in 578.42918ms (578.46711ms including waiting)
  Normal   Pulling    30s (x4 over 77s)  kubelet            Pulling image "quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec"
  Normal   Created    29s (x4 over 77s)  kubelet            Created container cloud-api-adaptor-con
  Warning  Failed     29s (x4 over 77s)  kubelet            Error: failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: eaccess /run/cloud-api-adaptor/plugins/entrypoint.sh: permission denied: unknown
  Normal   Pulled     29s                kubelet            Successfully pulled image "quay.io/stevenhorsman/cloud-api-adaptor:dev-4830f38ebfa4051281bad3ac9e318013658ff4ec" in 592.613617ms (592.65714ms including waiting)
  Warning  BackOff    17s (x9 over 75s)  kubelet            Back-off restarting failed container cloud-api-adaptor-con in pod cloud-api-adaptor-daemonset-js6rh_confidential-containers-system(c8afcf00-6476-442e-b82a-fe4b83cf8d95)
```

It's probably user error, but do you have an idea of what I might have done wrong?

Reply:
Double confirmed that the `/run/cloud-api-adaptor/plugins` folder do not work as expected.
Use the `/opt/cloud-api-adaptor/plugins` to replace it in the worker node.
And use the `/cloud-providers` as mount point inside container.

@stevenhorsman I update the document, please try again thanks.
And here is the images that I used base on the latest code in this PR.
- quay.io/liudalibj/peerpod-ctrl:latest
- quay.io/liudalibj/cloud-api-adaptor:dev-a8b81333ccb6b0e0adf71c8eda675da97d24d649

Reply:
the `/run` folder in worker node is be mounted with `noexec`
```
root@peer-pods-worker-0:~# mount |grep /run
tmpfs on /run type tmpfs (rw,nosuid,nodev,noexec,relatime,size=615648k,mode=755,inode64)
```

Reply:
With the updating documentation process the CAA seems to be working fine:
```
# kubectl logs -n confidential-containers-system ds/cloud-api-adaptor-daemonset
+ exec cloud-api-adaptor libvirt -uri 'qemu+ssh://root@192.168.122.1/system?no_verify=1' -data-dir /opt/data-dir -pods-dir /run/peerpod/pods -network-name default -pool-name default -disable-cvm -socket /run/peerpod/hypervisor.sock
2024/04/17 10:50:10 [adaptor/cloud] Loading external plugin libvirt from /cloud-providers/libvirt.so
2024/04/17 10:50:10 [adaptor/cloud] Successfully opened the external plugin /cloud-providers/libvirt.so
```
and the same for the peerpod-ctrl:
```
# kubectl logs peerpod-ctrl-controller-manager-85745445fd-zqq4p -n confidential-containers-system
2024-04-17T10:51:47Z	INFO	controller-runtime.metrics	Metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
2024/04/17 10:51:47 [adaptor/cloud] Loading external plugin libvirt from /cloud-providers/libvirt.so
2024/04/17 10:51:47 [adaptor/cloud] Successfully opened the external plugin /cloud-providers/libvirt.so
2024/04/17 10:51:47 [adaptor/cloud/libvirt] libvirt config: &libvirt.Config{URI:"qemu+ssh://root@192.168.122.1/system?no_verify=1", PoolName:"default", NetworkName:"default", DataDir:"", DisableCVM:false, VolName:"podvm-base.qcow2", LaunchSecurity:"", Firmware:"/usr/share/edk2/ovmf/OVMF_CODE.fd"}
2024/04/17 10:51:48 [adaptor/cloud/libvirt] Created libvirt connection
```